riscv-peripheral/src/hal_async/aclint.rs

```diff
@@ -0,0 +1,49 @@
+//! Asynchronous delay implementation for the (A)CLINT peripheral.
+
+use crate::aclint::mtimer::MTIME;
+pub use crate::hal::aclint::Delay;
+pub use crate::hal_async::delay::DelayUs;
+use core::{
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+struct DelayAsync {
+    mtime: MTIME,
+    t0: u64,
+    n_ticks: u64,
+}
+
+impl DelayAsync {
+    pub fn new(mtime: MTIME, n_ticks: u64) -> Self {
+        let t0 = mtime.read();
+        Self { mtime, t0, n_ticks }
+    }
+}
+
+impl Future for DelayAsync {
+    type Output = ();
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.mtime.read().wrapping_sub(self.t0) < self.n_ticks {
+            true => Poll::Pending,
```

Comment:
on `Pending` you should store the waker somewhere, and wake it when the timer expires. Otherwise the future will hang (except with bery dumb `block_on`-like executors that poll in a hot loop, not waiting for wakes)

Reply:
Thanks for the comment! I left asynchronous stuff for later once Rust 1.75 was out. I'll address it ASAP

Reply:
Hey, @Dirbaio I address your comments.

I'm pretty new to asynchronous programming, but if I'm not wrong this implementation is just a *soft* polling, as the executor must poll from time to time the `Future`. I was wondering about how an interrupt-driven approach would look like.

I guess that I would need a `static` placeholder to store all the pending contexts and wake them when the `MachineTimer` interrupt is triggered. However, it seems quite difficult to dimension this, as I don't know the number of contexts that may live concurrently in an asynchronous application. Can you send me a few pointers to get more insights about this?

Reply:
So... `async` stuff was more scary than I expected 😮‍💨 

My latest push provides a few `extern "Rust"` functions that  must  be implemented for pushing/popping `Timer`s to a timer queue. If I'm not wrong, my approach would be compatible with embassy-time or any other interface by properly implementing the `_riscv_peripheral*` functions.